src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java

```diff
@@ -75,16 +103,166 @@ private static FileSystem createFileSystem(URI uri, Configuration conf)
+    {
+        AuthenticationMethod authenticationMethod = userGroupInformation.getAuthenticationMethod();
+        switch (authenticationMethod) {
+            case SIMPLE:
+                return ImmutableSet.of();
+            case KERBEROS:
+                return ImmutableSet.copyOf(getSubject(userGroupInformation).getPrivateCredentials());
+            case PROXY:
+                return getPrivateCredentials(userGroupInformation.getRealUser());
+            default:
+                throw new IllegalArgumentException("Unsupported authentication method: " + authenticationMethod);
+        }
+    }
+
+    private static class FileSystemKey
+    {
+        private final String scheme;
+        private final String authority;
+        private final long unique;
+        private final String realUser;
+        private final String proxyUser;
+
+        private FileSystemKey(String scheme, String authority, long unique, String realUser, String proxyUser)
+        {
+            this.scheme = requireNonNull(scheme, "scheme is null");
+            this.authority = requireNonNull(authority, "authority is null");
+            this.unique = unique;
+            this.realUser = requireNonNull(realUser, "realUser");
+            this.proxyUser = proxyUser;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            FileSystemKey that = (FileSystemKey) o;
+
+            return Objects.equals(this.scheme, that.scheme) &&
+                    Objects.equals(this.authority, that.authority) &&
+                    Objects.equals(this.unique, that.unique) &&
+                    Objects.equals(this.realUser, that.realUser) &&
+                    Objects.equals(this.proxyUser, that.proxyUser);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(scheme, authority, unique, realUser, proxyUser);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("scheme", scheme)
+                    .add("authority", authority)
+                    .add("unique", unique)
+                    .add("realUser", realUser)
+                    .add("proxyUser", proxyUser)
+                    .toString();
+        }
+    }
+
+    private static class FileSystemHolder
     {
-        try {
-            Field field = FileSystem.Cache.class.getDeclaredField("map");
-            field.setAccessible(true);
-            return (Map<Key, FileSystem>) field.get(cache);
+        private final FileSystem fileSystem;
+        private final Set<?> privateCredentials;
+
+        private FileSystemHolder(FileSystem fileSystem, Set<?> privateCredentials)
```

Comment:
Nit: this should be public, since it's being called outside the class. (even though it technically works like this, public makes it more clear)